container-core/src/main/java/com/yahoo/processing/rendering/AsynchronousSectionedRenderer.java

```diff
@@ -334,26 +336,46 @@ private class DataListListener extends RendererListener {
         private final DataListListener parent;
 
+        /** Queue of rendering tasks that can be executed immediately without dispatching to executor and incuring a context switch */
+        private final Queue<Runnable> syncTasks = new LinkedList<>();
+
+
         public DataListListener(DataList list, DataListListener parent) {
             this.list = list;
             this.parent = parent;
         }
 
         @Override
         protected void render() throws IOException, InterruptedException, ExecutionException {
-            if (dataListListenerStack.peekFirst() != this)
-                return; // This listens to some ancestor of the current list, do this later
-            if (beforeHandoverMode &&  ! list.isFrozen())
-                return; // Called on completion of a list which is not frozen yet - hold off until frozen
-
-            if ( ! beforeHandoverMode)
-                list.completeFuture().get(); // trigger completion if not done already to invoke any listeners on that event
-            boolean startedRendering = renderData();
-            if ( ! startedRendering || uncompletedChildren > 0) return; // children must render to completion first
-            if (list.completeFuture().isDone()) // might not be when in before handover mode
-                endListLevel();
-            else
-                stream.flush();
+            try {
+                if (dataListListenerStack.peekFirst() != this)
+                    return; // This listens to some ancestor of the current list, do this later
+                if (beforeHandoverMode &&  ! list.isFrozen())
+                    return; // Called on completion of a list which is not frozen yet - hold off until frozen
+
+                if ( ! beforeHandoverMode)
+                    // Trigger completion if not done already to invoke any listeners on that event.
+                    // Note that the completable future might have its get() method overridden.
+                    // See DrainOnGetFuture.get() and friends for details.
+                    list.completeFuture().get();
+                boolean startedRendering = renderData();
+                if ( ! startedRendering || uncompletedChildren > 0) return; // children must render to completion first
+                if (list.completeFuture().isDone()) // might not be when in before handover mode
+                    endListLevel();
+                else
+                    stream.flush();
+            } finally {
+                RuntimeException exception = null;
+                while (!syncTasks.isEmpty()) {
+                    try {
+                        syncTasks.poll().run();
+                    } catch (RuntimeException e) {
+                        if (exception == null) exception = e;
+                        else exception.addSuppressed(e);
+                    }
+                }
+                if (exception != null) throw exception;
```

Comment:
This will hide exceptions thrown in the `try { }` block. Consider re-throwing those, with any additional exceptions here as suppressed.

Reply:
The exception will only be hidden if `finally` throws a new exception right? That is acceptable imho.

Reply:
`try-with-resources` automatically handles this, btw., when the `close()` there throws; then, that is added as a suppressed exception to whatever was thrown inside the `try { }` block; or it is thrown as the main exception if nothing else was thrown.